### ISSUE TITLE
embedly-proxy receiving null URLs

### ISSUE THREAD

**Issue by jaredkerim** [mozilla/activity-stream#518]
The activity stream addon is sending requests to the embedly-proxy client with empty URLs like so:
	
{"urls":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null]}

Each request has a variable number of URLs in the list, but are always all null.

I suspect this is coming directly from deployed addons (rather than network calls being inadvertently made during automated testing) because I see a distribution of user agents which looks like actual users.

**Comment by pdehaan** [comment]
Too bad we aren't passing the add-on version to the proxy to see if its all older add-ons.

**Comment by jaredkerim** [comment]
It's possible this is coming from automated tests but sentry reports that this is coming from multiple versions of Firefox and from multiple operating systems including OSX, Windows 7-10, and Linux.

**Comment by oyiptong** [comment]
oh interesting. it sounds like it's a deeper issue than just an automated testing config problem

**Comment by pdehaan** [comment]
@jaredkerim Are you still seeing this nullery in the logs?

I wasn't able to reproduce https://github.com/mozilla/embedly-proxy/issues/70 earlier today w/ testpilot.firefox.com, so wondering if it was just a transient error or solar flares or something.

**Comment by rlr** [comment]
I found the issue here. Or one of the issues that causes this. On startup, `ActivityStreams.js` builds a preview cache `this._asyncBuildPreviewCache();` but this function is missing one step. It calls `this._previewProvider.asyncSaveLinks(links);` without calling `links = this._previewProvider.processLinks(links);`. `processLinks()` is what sets the `sanitizedUrl` to pass to embedly among other things. Since it isn't called, it is passing null instead.

Want me to grab this @jaredkerim ?

**Comment by rlr** [comment]
This also means that the previewCache we create on startup isn't really working so fixing this might give a perf boost in some cases.

**Comment by rlr** [comment]
This should fix it... The hard part is writing the tests 😆 

```
diff --git a/lib/ActivityStreams.js b/lib/ActivityStreams.js
index 0c79f6a..268097b 100644
--- a/lib/ActivityStreams.js
+++ b/lib/ActivityStreams.js
@@ -342,6 +342,7 @@ ActivityStreams.prototype = {
       }));

       yield Promise.all(promises);
+      linksToSend = this._previewProvider.processLinks(linksToSend);
       yield this._previewProvider.asyncSaveLinks(linksToSend);
       this._populatingCache.preview = false;
       Services.obs.notifyObservers(null, "activity-streams-previews-cache-complete", null);
```

**Comment by rlr** [comment]
(At least it fixes it for me locally)

**Comment by sarracini** [comment]
For me even with @rlr 's fix, if you set the timeouts to something small to observe the calls it's making, I still get a bunch of null requests. But you're right, we should definitely be processing the links before priming the cache. I'm not sure that's enough to fix the null problem, it seems deeper than this

**Comment by sarracini** [comment]
More specifically this still reproduces the nulls:

````js
diff --git a/lib/PreviewProvider.js b/lib/PreviewProvider.js
index 0b53976..8946530 100644
--- a/lib/PreviewProvider.js
+++ b/lib/PreviewProvider.js
@@ -30,9 +30,9 @@ Cu.import("resource://gre/modules/Services.jsm");

 const DEFAULT_OPTIONS = {
   cacheCleanupPeriod: 86400000, // a cache clearing job runs at most once every 24 hours
-  cacheRefreshAge: 259200000, // refresh a link every 3 days
+  cacheRefreshAge: 10000, // refresh a link every 3 days
   cacheTTL: 2592000000, // cached items expire if they haven't been accessed in 30 days
-  cacheUpdateInterval: 3600000, // a cache update job runs every hour
+  cacheUpdateInterval: 10000, // a cache update job runs every hour
   proxyMaxLinks: 25, // number of links embedly proxy accepts per request
   initFresh: false,
 };
````

If you open the browser console's network tab and observe, every 10s you'll still get the nulls (even with ricky's patch).

**Comment by rlr** [comment]
Ok. For me it fixed the case of precaching on startup. I'll play with these timeouts too.

**Comment by sarracini** [comment]
I think I fixed it:

````js 
@@ -243,7 +243,8 @@ PreviewProvider.prototype = {
         links.push(link);
       }
     }
-    yield this.asyncSaveLinks(links, false, false);
+    let linksToUpdate = this.processLinks(links);
+    yield this.asyncSaveLinks(linksToUpdate, false, false);
     Services.obs.notifyObservers(null, "activity-streams-preview-cache-update", null);
   }),
````

I don't get nulls anymore with reduced timeouts with this fix

**Comment by rlr** [comment]
yay!

**Comment by sarracini** [comment]
fixed with #697
